### DISCUSSION TITLE
What's the difference between the distance and similarity functions?

### DISCUSSION THREAD

**Discussion by stijnfrishert** [maxbachmann/rapidfuzz-cpp#111] [Q&A] [answered]
What's the difference between functions such as `levenshtein_distance()`, `levenshtein_similarity()` and the normalized variants?

In [Levenshtein.hpp](https://github.com/maxbachmann/rapidfuzz-cpp/blob/b86b31b0575c996d6d85c41d9823529581662a46/rapidfuzz/distance/Levenshtein.hpp) there's two big comments.

- The [first comment block](https://github.com/maxbachmann/rapidfuzz-cpp/blob/b86b31b0575c996d6d85c41d9823529581662a46/rapidfuzz/distance/Levenshtein.hpp#L11) explains the Levenshtein distance just fine
- `levenshtein_similarity()` has no comment
- `levenshtein_normalized_distance()` has no comment
- The [second comment block](https://github.com/maxbachmann/rapidfuzz-cpp/blob/b86b31b0575c996d6d85c41d9823529581662a46/rapidfuzz/distance/Levenshtein.hpp#L191) purports to explain the `normalized_levenshtein()`, but is placed in front of `levenshtein_normalized_similarity()`?

I'm guessing similarity is just the opposite of distance (so a similarity of max distance means a full match), but that's guess work on my side. Maybe the comments could use a refresher. :)

**Comment by maxbachmann** [comment, marked as answer]
I never really got around to making the C++ documentation nice :sweat_smile: 

Your interpretation is correct about what these are:
- distance -> what you would regularly call Levenshtein distance, which is a score between 0 and max, where max for the uniform Levenshtein distance is max(len1, len2)
- similarity -> the inversed version, so a score between max and 0 where max is the optimum
- normalized_distance / normalized_similarity -> normalized version of the versions above to a range from 0.0-1.0 or 1.0-0.0 respectively

**Reply by stijnfrishert** [reply]
Makes a lot of sense, and thank you for clearing that up :)

Aside from that, very happy with the library!
